### DISCUSSION TITLE
Glyphwise for rtl text

### DISCUSSION THREAD

**Discussion by bensapirstein** [coldtype/coldtype#147] [Q&A]
I'm trying to apply Glyphwise on a RTL text and find that it doesn't write it in the right order, it replaces the first and last characters with the ones that follow and precede them, respectively.

```python
from coldtype import *

fnt = Font("IBM Plex/IBM plex master IBM-Plex-Sans-Hebrew-sources_masters/IBM Plex Sans Hebrew.designspace")
txt = "דיר\nבאלכם"

@animation((1080, 1080), timeline=Timeline(30, 30), bg=0)
def wave(f):
    return (Glyphwise(txt, lambda g:
                      Style(fnt, 200, wght=f.adj(-g.i*10).e("seio", 1, rng=(0.3, 0.7))))
        .xalign(f.a.r, "E")
        .align(f.a.r)
        .f(1))
```

https://github.com/coldtype/coldtype/assets/14977606/6342c8fa-1983-4d91-b24e-e26811a6a5f4

**Comment by stenson** [comment]
Definitely a good suggestion — Glyphwise at the moment is very limited in scope to LTR (and it's probably more limited than that), but this would be a good thing to add to the library, made an issue for it here: https://github.com/coldtype/coldtype/issues/148

**Comment by HinTak** [comment]
Shameless plug - https://github.com/kyamagu/skia-python/issues/195#issuecomment-1801101394 - the next release of skia-python can do rtl text somewhat automatically, with a newly added API.

**Comment by bensapirstein** [comment]
I managed to find a workaround, processing the string by reversing each word and padding it with 2 spaces. The spaces solves the bug that duplicates and erases margin letters, and reversing adjusts the text for RTL.

```python
from coldtype import *

def process(words):
    return '\n'.join('  ' + word[::-1] + '  ' for word in words.split())

fnt = Font("IBM Plex/IBM plex master IBM-Plex-Sans-Hebrew-sources_masters/IBM Plex Sans Hebrew.designspace")
txt = process("דיר\nבאלכם")

@animation((1080, 1080), timeline=Timeline(30, 30), bg=0)
def wave(f):
    return (Glyphwise(txt, lambda g:
                      Style(fnt, 200,
                            wght=f.adj(g.i*4).e("seio")))
        .xalign(f.a.r, "E")
        .align(f.a.r)
        .f(1))
``` 

https://github.com/coldtype/coldtype/assets/14977606/d70c1c50-ecfb-4f99-8fd7-e5fb6ca6218e

**Reply by bensapirstein** [reply]
Doesn't seem to work for Arabic

**Reply by HinTak** [reply]
I don't know how coldtype uses skia python, but that particular new part of newer skia python was explicitly tested for Arabic, I seem to remember. FYI.

**Reply by stenson** [reply]
Coldtype doesn't use skia-python at all for typographic support (it uses [uharfbuzz](https://github.com/harfbuzz/uharfbuzz) directly to expose glyph-level information about harfbuzz-shaped text) and it does support Arabic and other RTL scripts, this is just a thread about the `Glyphwise` class in Coldtype, which is an experimental way to set text with variable fonts and independent glyph-by-glyph styling, which is not officially supported by any typographic shaping algorithms (because spacing & kerning & contextual opentype features do not hold across variable font variations) — it could be useful to skia-python for some text support, but for this specific use case, the idea is to get information about the shaped position of each glyph in a run, rather than a single vector. All that said, I'm still not sure how Glyphwise could be adapted to RTL scripts without a complete rewrite, since it makes a lot of assumptions about the input text in order to get approximations of kerning and other features.

**Reply by HinTak** [reply]
Skia's shaper class uses harfbuzz on linux but coretext on mac - I think I already filed a bug that skia doesn’t use uniscribe on windows (or anything) so skia's shaper class is buggy on windows. That said, what you are asking for - individual positionings of glyph runs, etc, seem to be possible with skia - https://github.com/google/skia/blob/91ddbf06eaa3b1343c89c31a313fe9b5268fdab1/modules/skshaper/include/SkShaper.h#L216 - I just haven't binded that in skia-python yet, as it seems a lot of work for very niche use. If you think having access to that part of skia is useful, we can think about it.
